File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int low = 0, high = 0, size = nums.size();
        multiset<int> currentWindow;
        vector<int> result;
        while(high < size){
            if(high - low + 1 < k){
                currentWindow.insert(nums[high]);
                high++;  
            }
            else{
                currentWindow.insert(nums[high]);
                auto maxNum = currentWindow.rbegin();
                result.push_back(*maxNum);
                currentWindow.erase(currentWindow.find(nums[low]));
                low++;
                high++;
            }
        }
        return result;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        q = collections.deque()

        for i in range(0,k):
            if q: 
                while q and nums[q[-1]] < nums[i]: q.pop()
            q.append(i)

        res.append(nums[q[0]])

        l, r = 1, k
        while r < len(nums):
            if (l - 1) == q[0]: q.popleft()
            while q and nums[q[-1]] < nums[r]: q.pop()
            q.append(r)
            r += 1
            l += 1
            res.append(nums[q[0]])
        
        return res
            


